app/src/main/java/com/hedvig/app/feature/home/ui/changeaddress/GetAddressChangeStoryIdUseCase.kt

```diff
@@ -1,28 +1,49 @@
 package com.hedvig.app.feature.home.ui.changeaddress
 
+import arrow.core.Either
+import arrow.core.flatMap
 import com.apollographql.apollo.ApolloClient
 import com.hedvig.android.owldroid.graphql.ActiveContractBundlesQuery
-import com.hedvig.app.feature.home.ui.changeaddress.GetAddressChangeStoryIdUseCase.SelfChangeEligibilityResult.Blocked
-import com.hedvig.app.feature.home.ui.changeaddress.GetAddressChangeStoryIdUseCase.SelfChangeEligibilityResult.Eligible
-import com.hedvig.app.feature.home.ui.changeaddress.GetAddressChangeStoryIdUseCase.SelfChangeEligibilityResult.Error
+import com.hedvig.app.feature.embark.QUOTE_CART_ID_KEY
+import com.hedvig.app.feature.embark.quotecart.CreateQuoteCartUseCase
+import com.hedvig.app.util.ErrorMessage
 import com.hedvig.app.util.apollo.QueryResult
 import com.hedvig.app.util.apollo.safeQuery
 
 class GetAddressChangeStoryIdUseCase(
+    private val createQuoteCartUseCase: CreateQuoteCartUseCase,
     private val apolloClient: ApolloClient,
 ) {
 
     suspend operator fun invoke(): SelfChangeEligibilityResult {
-        return when (val result = apolloClient.query(ActiveContractBundlesQuery()).safeQuery()) {
-            is QueryResult.Success ->
-                result.data
+        return (
+            apolloClient.query(ActiveContractBundlesQuery()).safeQuery().toEither().flatMap { data ->
+                val storyId = data
                     .activeContractBundles
                     .firstOrNull()
                     ?.angelStories
                     ?.addressChangeV2
-                    ?.let(::Eligible)
-                    ?: Blocked
-            is QueryResult.Error -> Error(result.message)
+                if (storyId != null) {
+                    addQuoteCartId(storyId).mapLeft { QueryResult.Error.GeneralError(it.message) }
+                } else {
+                    Either.Right(null)
```

Comment:
I think we're actually using `Right` for the happy path and `Left` for the unhappy path here 🤔

Reply:
We're using Right for the case where the storyId is null and later in the fold case in the "Right" we have to do a check if we're on the happy path (Eligible) or if it's null we're on the unhappy path (Blocked) while being inside the `ifRight` which is like an extra mental overhead having to also care for "bad" cases while on the right path.

Reply:
I think this might be a bit philosophical, but the `Left`-path represents the unknown-types of Errors here (network errors and so on), while the `Right`-path represents the known errors, if you will - it's well-known that you may not be allowed to perform an address change, without there being any errors involved (this is the `Blocked`-case).

Reply:
Aha that's your point of view, okay I get it so I can better discuss it now.

The way I see it, Left is the cases where you didn't get the "successful" result, not so much about being <Unknown errors, Known Failed result | Success result> inside the Either. But I prefer <Unknown errors | Known Failed Result , Successful result>.
Otherwise we'd also have to change usecases like [this one](https://github.com/HedvigInsurance/android/blob/e8031451750b3dce3b8bc868a5784de0b3b6275d/app/src/main/java/com/hedvig/app/feature/insurance/ui/detail/GetContractDetailsUseCase.kt#L21) to return <NetworkError, ContractNotFound | ContractDetailViewState> instead of <NetworkError|ContractNotFound, ContractDetailViewState> Since the contract not found is a known result that we know we can get back.

Or to put it in another way, the use case "GetAddressChangeStoryIdUseCase" returns the address change story Id or a sealed union of error cases the same way that the use case "GetContractDetailsUseCase" returns the contract details or a sealed union of error cases. Whether we exactly what went wrong or if it's a generic error that we don't know exactly what happened is just a matter of adding more union types to the Left side.

And the Eithers that are handling the logic inside the UseCase should map 1:1 to what the UseCase is gonna return to make it easier mentally for everyone to see what's going on instead of then having to take the left and turn it into the error path, and take the right and check if it's the actual right path of it's the right path that's actually an error.

That's how I view this and structure my Eithers at least. I'd love to hear @hugokallstrom opinion on this too actually.

Reply:
The way I see it, `Either` replaces the use of Exceptions and places them on the `Left`-field, and Exceptions mostly represents unknown errors - the use of exceptions for known errors, except for checked exceptions as in Java, is mostly considered bad practice (afaik).

I suppose if we want to use `Either` for more flow-control purposes, having `Left` represent any type of 'negative' case could be a thing. I'd also like to hear @hugokallstrom opinion here.